trillian/integration/chains.go

```diff
@@ -0,0 +1,245 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package integration
+
+import (
+	"crypto"
+	"crypto/rand"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/google/certificate-transparency-go/x509"
+	"github.com/google/trillian/crypto/keys"
+	"github.com/google/trillian/crypto/keyspb"
+
+	ct "github.com/google/certificate-transparency-go"
+)
+
+// ChainGenerator encapsulates objects that can generate certificate chains for testing.
+type ChainGenerator interface {
+	// CertChain generates a certificate chain.
+	CertChain() ([]ct.ASN1Cert, error)
+	// PreCertChain generates a precertificate chain, and also returns the leaf TBS data
+	PreCertChain() ([]ct.ASN1Cert, []byte, error)
+}
+
+// SyntheticChainGenerator builds synthetic certificate chains based on
+// a template chain and intermediate CA private key.
+type SyntheticChainGenerator struct {
+	chain    []ct.ASN1Cert
+	leafCert *x509.Certificate
+	caCert   *x509.Certificate
+	// Signer which matches the caCert
+	signer   crypto.Signer
+	notAfter time.Time
+}
+
+// NewSyntheticChainGenerator returns a ChainGenerator that mints synthetic certificates based on the
+// given template chain.  The provided signer should match the public key of the first issuer cert.
+func NewSyntheticChainGenerator(chain []ct.ASN1Cert, signer crypto.Signer, notAfter time.Time) (ChainGenerator, error) {
+	if len(chain) < 2 {
+		return nil, fmt.Errorf("chain too short (%d)", len(chain))
+	}
+	leaf, err := x509.ParseCertificate(chain[0].Data)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse leaf cert: %v", err)
+	}
+	issuer, err := x509.ParseCertificate(chain[1].Data)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse issuer cert: %v", err)
+	}
+	if notAfter.IsZero() {
+		notAfter = time.Now().Add(24 * time.Hour)
```

Comment:
If this was required to be non nil then this level of code wouldn't depend on the current time.

Reply:
Yeah, but then the same few lines of code to enforce non-zeroness would need to be repeated in 3 of the 4 callers.

Reply:
Probably not a big issue for test code.